### PR TITLE
Configurable losses

### PR DESCRIPTION
**Pull request by Zhylkaaa**
this pr adds 3 losses in addition to CrossEntropy. 
Also refactors loss calculation extracting it to separate file (as it's basically the same for classification and ner) 
body of `_create_training_progress_scores` has been replaced with defaultdict.
Also optional evaluation od test data has been added.

### CODE REVIEW

**Comment by Zhylkaaa** [comment]
I have used focal loss and got good results, but other 2 seem to only make things worse. 
https://arxiv.org/pdf/1911.02855.pdf - this publication states that dice loss can help training, so maybe this is something wrong with my implementations. Can somebody tak a look?

**Comment by Zhylkaaa** [comment]
I have changed loss calculation according to your review in #1208 and refactored tensor board logging to much wandb logging (without custom plots from wandb)

**Comment by Zhylkaaa** [comment]
Also where was a bug with installation, transformers 4.2 don't have `BigBird` model, so I have changed it to version 4.6

**Comment by ZYG151541155** [comment]
Hi,  how can I change classification loss function from crossentropy to focal loss or dice loss? Can you give me some tips or help me to solve this question?

**Comment by Zhylkaaa** [comment]
hello, @ZYG151541155, you should install package from this branch and then specify something like this:
You can also read docstrings in losses/focal_loss.py to find what this arguments do

```
model_args.loss_type = 'focal'
model_args.loss_args = {
            'alpha': None,
            'gamma': 2,
            'reduction': 'mean',
            'ignore_index': -100,
        }
```

**Comment by ZYG151541155** [comment]
> hello, @ZYG151541155, you should install package from this branch and then specify something like this:
> You can also read docstrings in losses/focal_loss.py to find what this arguments do
> 
> ```
> model_args.loss_type = 'focal' # or 'dice' or 'tversky'
> model_args.loss_args = {
>             'alpha': None,
>             'gamma': 2,
>             'reduction': 'mean',
>             'ignore_index': -100,
>         }
> ```
> 
> EDIT: actually I am not sure mu dice loss implementation is good, but maybe I have used bad parameters 🤷‍♂️
> If what I've said doesn't work please comment and paste the error message hear
> (you can also contact me via mail in my profile)

Ohhh，thanks a lot. Can you tell me this branch version? I don't know which branch should I download:
image
After check out my data result , I will record score and put it in here.

**Comment by Zhylkaaa** [comment]
If you are using `pip` just run `python -m pip install git+https://github.com/Zhylkaaa/simpletransformers.git@add_losses` 
Which will install library from `add_losses` branch from my fork (I am not a maintainer, so I can't create branches in main repo)

**Comment by Zhylkaaa** [comment]
If somebody is interested in looking into dice loss you can start hear https://github.com/ShannonAI/dice_loss_for_NLP/issues/11

**Comment by ZYG151541155** [comment]
> If you are using `pip` just run `python -m pip install git+https://github.com/Zhylkaaa/simpletransformers.git@add_losses`
> Which will install library from `add_losses` branch from my fork (I am not a maintainer, so I can't create branches in main repo)

Ok, Thanks bro. And I have other question, 

> If you are using `pip` just run `python -m pip install git+https://github.com/Zhylkaaa/simpletransformers.git@add_losses`
> Which will install library from `add_losses` branch from my fork (I am not a maintainer, so I can't create branches in main repo)

Ohhh，thanks my bro.

**Comment by Zhylkaaa** [comment]
It may be also helpful if one could pass lambda as a loss function, but I am not sure about how to implement this elegantly. Do someone have any idea?

**Comment by Zhylkaaa** [comment]
@ThilinaRajapakse should I prepare some documentation or examples for this?

**Comment by ThilinaRajapakse** [comment]
I think this is good to go except for one change I made. I think it's better to keep the `_calculate_loss()` function as part of the model itself. The loss calculation is different for classification vs NER and I don't think this was being accounted for in the common `_calculate_loss()` function. So I just kept the old functions which were part of the classification model and NER model.

**Comment by ThilinaRajapakse** [comment]
> @ThilinaRajapakse should I prepare some documentation or examples for this?

That would be great! I just made the change I mentioned and merged it. Let me know if that works or not.

**Comment by Zhylkaaa** [comment]
ok, I actually updated dice loss and tversky loss but haven't pushed it yet, so I will add this soon and add some documentation on what this parameters do

**Comment by ThilinaRajapakse** [comment]
Thank you for your hard work!

**Comment by ThilinaRajapakse** [comment]
> It may be also helpful if one could pass lambda as a loss function, but I am not sure about how to implement this elegantly. Do someone have any idea?

I think we could just add `"custom"` as an option for `model_args.loss_type` and `loss_fct` optional parameter to the init function of `ClassificationModel` and `NERModel`. If `model_args.loss_type == "custom"`, then `loss_fct` should be provided.

`loss_fct` should accept the model outputs and labels and return the loss.

**Comment by joachimdb** [comment]
Hi all, has this feature been documented already? I'm trying to use a custom loss, but I'm not sure how to do that. Specifically, when I specify a bad value for the `loss_type` argument, I would expect that an exception is raised, but it isn't?

My code:
```
train_df = [...]
args = MultiLabelClassificationArgs(
    [...]
    loss_type = "ml_an_ls".  # <-- there is no implementation for this loss_type
)

m = MultiLabelClassificationModel(
    'xlmroberta',
    'xlm-roberta-base',
    [...]
    args=args)

m.train_model(train_df).  # <- but this does not raise an error?
```

**Comment by joachimdb** [comment]
Update: I figured out what is going on: the custom loss functionality is not implemented for the multi-label case. Is there a particular reason for that (or any plans for implementing it?)

**Comment by shibing624** [comment]
please add focal loss for sigmoid with SIGMOID_FOCAL_CROSS_ENTROPY.
